consensus/hotstuff/votecollector/factory.go

```diff
@@ -0,0 +1,114 @@
+package votecollector
+
+import (
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/consensus/hotstuff/signature"
+	"github.com/onflow/flow-go/consensus/hotstuff/verification"
+	"github.com/onflow/flow-go/model/encoding"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+)
+
+// VoteProcessorFactory is a proxy class that delegates construction of vote processor to
+// base factory. Base factory is responsible only for creating objects of combined or staking
+// vote processors while VoteProcessorFactory is responsible for processing proposer vote and
+// returning constructed object.
+type VoteProcessorFactory struct {
+	base hotstuff.VoteProcessorFactory
+}
+
+// CombinedBaseVoteProcessorFactory implements a factory for creating CombinedVoteProcessor
+// holds needed dependencies to initialize CombinedVoteProcessor.
+type CombinedBaseVoteProcessorFactory struct {
+	log         zerolog.Logger
+	packer      hotstuff.Packer
+	committee   hotstuff.Committee
+	onQCCreated hotstuff.OnQCCreated
+}
+
+var _ hotstuff.VoteProcessorFactory = &CombinedBaseVoteProcessorFactory{}
+var _ hotstuff.VoteProcessorFactory = &VoteProcessorFactory{}
+
+func NewCombinedVoteProcessorFactory(log zerolog.Logger, committee hotstuff.Committee, eventHandler hotstuff.EventHandlerV2) *VoteProcessorFactory {
+	return &VoteProcessorFactory{
+		base: &CombinedBaseVoteProcessorFactory{
+			log:       log,
+			committee: committee,
+			packer:    signature.NewConsensusSigDataPacker(committee),
+			onQCCreated: func(qc *flow.QuorumCertificate) {
+				err := eventHandler.OnQCConstructed(qc)
```

Comment:
⚠️  We need to give the QC to event loop, instead of EventHandler. 

Note multiple QC will be created by the `onQCCreated` called concurrently, and EventHandler can't process them concurrently. So we need to give it to EventLoop.

Reply:
good catch

Reply:
💯 Thanks for spotting this, Leo!

Reply:
We could add a worker queue for event loop, the same way we do with engines, just with one worker, this way we won't block anyone and keep processing in order. 
Resolution of this comment depends on https://github.com/onflow/flow-go/pull/1437.